src/expr/dtype.cpp

```diff
@@ -167,7 +167,7 @@ bool DType::resolve(const std::map<std::string, TypeNode>& resolutions,
 
   d_involvesExt = false;
   d_involvesUt = false;
-  for (const std::shared_ptr<DTypeConstructor> ctor : d_constructors)
+  for (const std::shared_ptr<DTypeConstructor>& ctor : d_constructors)
```

Comment:
How about adding a reference in the loop before as well?
Also: why not `const auto&`?
```suggestion
  for (const auto& ctor : d_constructors)
```

Reply:
While I see your concern, I'd still push for using `auto`, at least if the container we are iterating has been declared within a few lines of the loop. (This is not the case here, but for almost all the remaining comments).
I argue that the inconvenience of having long text (that don't really further the understanding of what's going on) and the danger of getting the type wrong (maybe also forgetting to change it in a refactoring, giving rise to superfluous copying) outweighs the benefits of seeing the type right there instead of having to look for the declaration of the container, at least if said declaration is only a few lines away.

Reply:
I agree with @nafur and we explicitly allow the use of `auto` in our guidelines for exactly that reason. @ajreynol if you prefer to not use it , that's fine, but it is certainly allowed in general and even encouraged in some cases: https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#auto